source/base/parameter_handler.cc

```diff
@@ -737,6 +737,38 @@ namespace
           }
       }
   }
+
+  void
+  recursively_keep_non_default(const boost::property_tree::ptree &tree_in,
+                               boost::property_tree::ptree       &tree_out)
```

Comment:
Would you mind adding documentation to the function?

Reply:
I think the function is also misnamed. You're not "keeping" or "discarding" anything. You're *copying*.

I wonder if a more modern interface would just return a tree, rather than put it into the output argument? That would also simplify the use case below.